test/mocks/buffer/mocks.h

```diff
@@ -8,6 +8,49 @@
 #include "gmock/gmock.h"
 
 namespace Envoy {
+
+class MockBuffer : public Buffer::OwnedImpl {
+public:
+  MockBuffer() {
+    ON_CALL(*this, write(testing::_))
+        .WillByDefault(testing::Invoke(this, &MockBuffer::TrackWrites));
+    ON_CALL(*this, move(testing::_)).WillByDefault(testing::Invoke(this, &MockBuffer::BaseMove));
+  }
+
+  MOCK_METHOD1(write, int(int fd));
+  MOCK_METHOD1(move, void(Instance& rhs));
+  MOCK_METHOD2(move, void(Instance& rhs, uint64_t length));
+  MOCK_METHOD1(drain, void(uint64_t size));
+
+  void BaseMove(Instance& rhs) { Buffer::OwnedImpl::move(rhs); }
+  void BaseDrain(uint64_t size) { Buffer::OwnedImpl::drain(size); }
+
+  int TrackWrites(int fd) {
+    int bytes_written = Buffer::OwnedImpl::write(fd);
+    if (bytes_written > 0) {
+      bytes_written_ += bytes_written;
+    }
+    return bytes_written;
+  }
+
+  int FailWrite(int) {
```

Comment:
If it's meant for testing only the usual convention is to name the method something like failWriteForTest(). This doesn't appear to be in the style doc, I'll create a PR that mentions this convention.